### PR TITLE
#392 - Write photos taken via the app into the camera roll and update their location metadata

### PR DESCRIPTION
**Pull request by budowski**

### CODE REVIEW

File: src/components/ObsEdit/EvidenceSection.js [hunk 1]
```diff
@@ -91,6 +95,19 @@ const EvidenceSection = ( {
...
+      // We wrap this in a setTimeout since sometimes the location gets returned quickly, before
+      // the background code in ObsEditProvider finished writing the photos into the camera roll
+      // (and in this case, we don't mind updating the photos a little later, since it's all done
+      // in the background anyhow).
+      if ( location ) {
+        setTimeout( () => {
+          cameraRollUrisRef.current.forEach( uri => {
+            writeExifToFile( uri, { latitude: location.latitude, longitude: location.longitude } );
+          } );
+        }, 3000 );
```
Comment (kueda) [inline src/components/ObsEdit/EvidenceSection.js:108]: 1. I don't love using `setTimeout` here. If you need to wait until a file is written to a specific URI, why not track the the state of write success and only try to write the EXIF when you know the file exists? That would resolve the race condition in a more deterministic way
2. I don't think this component needs to know about `cameraRollUris`. Maybe ObsEditProvider should export a method like `updateCameraRollPhotosWithExif()` or something?

**Review by kueda — CHANGES_REQUESTED** [review]
I haven't done a very thorough check, but I've definitely found problems in both Android and iOS.

- [x] Please update package-lock.json with the new version of `react-native-exif-reader`
- [x] This often (but not always) results in multiple copies of the photo being written out for me on my Pixel 3, generally the first without coordinates and the rest with coordinates; it should only write a single photo out, preferably with coordinates
- [x] As far as I can tell it never writes EXIF location to the photo in iOS; I only have my iPad Air 5th gen running iPadOS 16.2 to test, though; I'm checking the Google Photos app and the Apple Photos app

**Comment by budowski** [comment]
> Ok, it seems to be writing coordinates out in iOS for me now, but it still writes out multiple photos in Android, usually after taking more than 1 photo. I tried to add some logging around this. Can you reproduce that, @budowski?

Weird, cannot recreate this on my PIxel 7 :(
Maybe it's a weird timing issue - can you send out logs from your device?

**Review by kueda — CHANGES_REQUESTED** [review]
Ok, Android problems with duplicate photos seem fixed! I'm still having a weird issue with iOS though: when I take a photo in the app and save the observation, I can see that photo in Apple Photos and Google Photos, but I can only see the location in Apple Photos. When I take a photo with the system camera, I can see the photo *and* the location in both Apple and Google Photos. Can you replicate that, and do you have any idea why that might be?

File: src/components/ObsEdit/EvidenceSection.js [hunk 2]
```diff
@@ -67,6 +72,21 @@ const EvidenceSection = ( {
...
+    }
+
+    // In theory we don't need to update any photos when this component is not mounted
+    if ( !mountedRef.current ) return;
+
+    // Update all photos taken via the app with the new fetched location.
+    cameraRollUris.forEach( uri => {
+      logger.debug( "calling writeExifToFile for uri: ", uri );
+      writeExifToFile( uri, { latitude, longitude, positional_accuracy: positionalAccuracy } );
+    } );
```
[the remark this answers is not in the document]
Comment (budowski) [reply src/components/ObsEdit/EvidenceSection.js]: Couldn't recreate this issue (the location that does show up in Google Photos). I did move the code to the new `writeExifToCameraRollPhotos` function

**Comment by kueda** [comment]
I've made one fairly significant change here: we're now writing EXIF when you *save* the observation, not every time we update the coordinates. This *seems* to prevent the duplicate photo issue in Google Photos on Android, which seemed like it might have been related to updating the photo's EXIF *while* Google Photos was importing the photo, because the photo never seemed to be duplicated in Files, only in Google Photos.

There still seems to be a problem when landscape photos in Android appearing upside down when written out. I think this is because the way we write them out does not preserve their `Orientation` metadata. `react-native-camera-roll` doesn't seem to support this, so we probably need to write the EXIF ourselves, but we can't do it until we [update `react-native-exif-reader` to write arbitrary EXIF tags](https://github.com/inaturalist/react-native-exif-reader/issues/4). If tests pass I'm going to merge despite that bug and deal with it later.

**Review by kueda — APPROVED** [review]

**Pull request merged**